pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,327 @@ package traceroute
+				var opErr error
+				if err := c.Control(func(fd uintptr) {
+					opErr = syscall.SetsockoptInt(int(fd), syscall.IPPROTO_IP, syscall.IP_TTL, ttl)
+				}); err != nil {
+					return err
+				}
+				return opErr
+			},
+		}
 
-type Target struct {
-	// The address of the target to traceroute to. Can be a DNS name or an IP address
-	Addr string `json:"addr" yaml:"addr" mapstructure:"addr"`
-	// The port to traceroute to
-	Port uint16 `json:"port" yaml:"port" mapstructure:"port"`
+		// Attempt to connect to the target host
+		conn, err := dialer.Dial("tcp", addr.String())
+		if !errors.Is(err, syscall.EADDRINUSE) {
+			return conn, port, err
+		}
+	}
 }
 
-func NewCheck() checks.Check {
-	return &Traceroute{
-		CheckBase: checks.CheckBase{
-			Mu:       sync.Mutex{},
-			DoneChan: make(chan struct{}),
-		},
-		config:     Config{},
-		traceroute: newTraceroute,
+// readIcmpMessage reads a packet from the provided icmp Connection. If the packet is 'Time Exceeded',
+// it reads the address of the router that dropped created the icmp packet. It also reads the source port
+// from the payload and finds the source port used by the previous tcp connection. If any error is returned,
+// an icmp packet was either not received, or the received packet was not a time exceeded.
+func readIcmpMessage(ctx context.Context, icmpListener *icmp.PacketConn, timeout time.Duration) (int, net.Addr, error) {
+	log := logger.FromContext(ctx)
+	// Expected to fail due to TTL expiry, listen for ICMP response
+	if err := icmpListener.SetReadDeadline(time.Now().Add(timeout)); err != nil {
+		return 0, nil, fmt.Errorf("failed to set icmp read deadline: %w", err)
+	}
+	buffer := make([]byte, 1500) //nolint:mnd // Standard MTU size
+	n, routerAddr, err := icmpListener.ReadFrom(buffer)
+	if err != nil {
+		// we probably timed out so return
+		return 0, nil, fmt.Errorf("failed to read from icmp connection: %w", err)
 	}
-}
 
-type Traceroute struct {
-	checks.CheckBase
-	config     Config
-	traceroute tracerouteFactory
-}
+	// Parse the ICMP message
+	msg, err := icmp.ParseMessage(ipv4.ICMPTypeTimeExceeded.Protocol(), buffer[:n])
+	if err != nil {
+		return 0, nil, err
+	}
 
-type tracerouteFactory func(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error)
+	// Ensure the message is an ICMP Time Exceeded message
+	if msg.Type != ipv4.ICMPTypeTimeExceeded {
+		log.Debug("message is not 'Time Exceeded'", "type", msg.Type.Protocol())
+		return 0, nil, errors.New("message is not 'Time Exceeded'")
+	}
 
-func newTraceroute(dest string, port, timeout, retries, maxHops int) (traceroute.TracerouteResult, error) {
-	opts := &traceroute.TracerouteOptions{}
-	opts.SetTimeoutMs(timeout)
-	opts.SetRetries(retries)
-	opts.SetMaxHops(maxHops)
-	opts.SetPort(port)
-	return traceroute.Traceroute(dest, opts)
-}
+	// The first 20 bytes of Data are the IP header, so the TCP segment starts at byte 20
+	tcpSegment := msg.Body.(*icmp.TimeExceeded).Data[20:]
```

Comment:
Consider adding a constant for this, so it's clearer when reading just the code.